File: frontend/src/lib/photoOperations.ts
/**
 * Photo Operations - Pure Business Logic
 */

import type { PhotoData, SourceConfig, Bounds } from './photoWorkerTypes';
import { PhotoSourceFactory, type PhotoSourceOptions } from './sources/PhotoSourceFactory';
import type { PhotoSourceLoader, PhotoSourceCallbacks } from './sources/PhotoSourceLoader';
import { filterPhotosByArea } from './workerUtils';
import { MAX_PHOTOS_IN_AREA } from './photoWorkerConstants';

// Import worker version for validation
declare const __WORKER_VERSION__: string;

export interface OperationCallbacks {
    shouldAbort: (processId: string) => boolean;
    postMessage: (message: any) => void;
    updatePhotosInArea: (photos: PhotoData[]) => void;
    updatePhotosInRange: (photos: PhotoData[]) => void;
    getPhotosInArea: () => PhotoData[];
    getPhotosInRange: () => PhotoData[];
    sendPhotosInAreaUpdate: () => void;
    sendPhotosInRangeUpdate: () => void;
    getValidToken: () => Promise<string | null>;
}

interface SourceCache {
    photos: PhotoData[];
    isComplete: boolean; // true = no more photos available, false = partial load (stream has "next" link)
    cachedBounds?: Bounds; // The geographic bounds that were completely cached
}

export class PhotoOperations {
    private loadingProcesses = new Map<string, PhotoSourceLoader>();
    private sourceCache = new Map<string, SourceCache>(); // Cache for each source
    private maxPhotosInArea: number = MAX_PHOTOS_IN_AREA;

    constructor() {}

    setMaxPhotosInArea(maxPhotos: number): void {
        this.maxPhotosInArea = maxPhotos;
    }

    /**
     * Clean up all resources - call this when worker is being terminated
     */
    cleanup(): void {
        console.log('🢄PhotoOperations: Cleaning up all resources');
        // Cancel all active loading processes
        for (const [sourceId, process] of this.loadingProcesses.entries()) {
            console.log(`🢄PhotoOperations: Cancelling loader for ${sourceId}`);
            process.cancel();
        }
        this.loadingProcesses.clear();

        // Clear all caches
        this.sourceCache.clear();
    }

    // Check if requestedBounds is completely contained within cachedBounds
    private isAreaWithinCachedBounds(requestedBounds: Bounds, cachedBounds: Bounds): boolean {
        return requestedBounds.top_left.lat <= cachedBounds.top_left.lat &&
               requestedBounds.top_left.lng >= cachedBounds.top_left.lng &&
               requestedBounds.bottom_right.lat >= cachedBounds.bottom_right.lat &&
               requestedBounds.bottom_right.lng <= cachedBounds.bottom_right.lng;
    }

    async processConfig(
        processId: string,
        messageId: number,
        config: { sources: SourceConfig[]; expectedWorkerVersion?: string; [key: string]: any },
        callbacks: OperationCallbacks
    ): Promise<void> {
        console.log(`🢄PhotoOperations: Processing config update (${processId})`);

        if (callbacks.shouldAbort(processId)) {
            console.log(`🢄PhotoOperations: PROCESSCONFIG: Config process ${processId} aborted before completion`);
            return;
        }

        // Check worker version if provided
        if (config.expectedWorkerVersion && config.expectedWorkerVersion !== __WORKER_VERSION__) {
            throw new Error(`Worker version mismatch! Expected: ${config.expectedWorkerVersion}, Actual: ${__WORKER_VERSION__}`);
        }

        if (!config?.sources) {
            throw new Error('Invalid config: missing sources array');
        }

        const sources = config.sources;
        let allLoadedPhotos: PhotoData[] = [];
        const enabledSourceIds = new Set(sources.filter(s => s.enabled).map(s => s.id));
        console.log(`🢄PhotoOperations: PROCESSCONFIG: enabledSourceIds: ${Array.from(enabledSourceIds).join(', ')}, loadingProcesses: ${Array.from(this.loadingProcesses.keys()).join(', ')}`);

        for (const [sourceId, process] of this.loadingProcesses.entries()) {
            if (!enabledSourceIds.has(sourceId)) {
                console.log(`🢄PhotoOperations: PROCESSCONFIG: Cancelling loading process for disabled source ${sourceId}`);
                process.cancel();
                this.loadingProcesses.delete(sourceId);
                // Also clear the cache for disabled sources to free memory
                //this.sourceCache.delete(sourceId);
                //console.log(`🢄PhotoOperations: PROCESSCONFIG: Cleared cache for disabled source ${sourceId}`);
            }
        }

        // Update photosInArea with all loaded photos
        // (Stream and device sources add their photos via callbacks)
        callbacks.updatePhotosInArea(allLoadedPhotos);
        callbacks.sendPhotosInAreaUpdate();

        console.log(`🢄PhotoOperations: PROCESSCONFIG: Config processing complete (${processId}) - loaded ${allLoadedPhotos.length} photos`);
        callbacks.postMessage({
            type: 'processComplete',
            processId,
            processType: 'config',
            messageId
        });
    }

    async processArea(
        processId: string,
        messageId: number,
        area: Bounds,
        sources: SourceConfig[],
        callbacks: OperationCallbacks
    ): Promise<void> {
        console.log(`🢄PhotoOperations: Processing area update (${processId}) with ${sources.length} sources`);

        if (callbacks.shouldAbort(processId)) return;

        if (!area) {
            console.log(`🢄PhotoOperations: No area bounds provided (${processId})`);
            callbacks.postMessage({
                type: 'processComplete',
                processId,
                processType: 'area',
                messageId
            });
            return;
        }

        // Area operations create FRESH arrays (never modify existing)
        let newPhotosInArea: PhotoData[] = [];

        // For each enabled source, collect photos in area
        if (sources) {
            for (const source of sources.filter(s => s.enabled)) {
                if (callbacks.shouldAbort(processId)) return;

                const cache = this.sourceCache.get(source.id);

                if (cache && cache.isComplete && cache.cachedBounds && this.isAreaWithinCachedBounds(area, cache.cachedBounds)) {
                    // Cache is complete AND current area is within cached bounds - use cache
                    console.log(`🢄PhotoOperations: Current area is within cached bounds for ${source.id}, filtering cached photos`);
                    const filteredPhotos = filterPhotosByArea(cache.photos, area);
                    console.log(`🢄PhotoOperations: Filtered ${filteredPhotos.length} photos from ${cache.photos.length} cached for ${source.id}`);

                    if (filteredPhotos.length > 0) {
                        newPhotosInArea.push(...filteredPhotos);
                    }
                } else if (cache && !cache.isComplete) {
                    // Cache is partial - need to perform bounded load
                    console.log(`🢄PhotoOperations: Cache for ${source.id} is partial, performing bounded load`);
                    await this.loadSource(source, processId, callbacks, area);
                } else {
                    // No cache - perform bounded load
                    console.log(`🢄PhotoOperations: No cache for ${source.id}, performing bounded load`);
                    await this.loadSource(source, processId, callbacks, area);
                }
            }
        }

        if (callbacks.shouldAbort(processId)) return;

        callbacks.updatePhotosInArea(newPhotosInArea);
        callbacks.sendPhotosInAreaUpdate();

        console.log(`🢄PhotoOperations: Area processing complete (${processId}) - ${newPhotosInArea.length} photos in area`);
        callbacks.postMessage({
            type: 'processComplete',
            processId,
            processType: 'area',
            messageId
        });
    }


    private async loadSource(
        source: SourceConfig,
        processId: string,
        callbacks: OperationCallbacks,
        bounds?: Bounds
    ): Promise<void> {
        // Cancel existing loader for this source if any
        const existingProcess = this.loadingProcesses.get(source.id);
        if (existingProcess) {
            console.log(`🢄PhotoOperations: Cancelling existing loader for source ${source.id} (process: ${processId})`);
            existingProcess.cancel();
        }

        // Create callbacks for all source types
        const sourceCallbacks: PhotoSourceCallbacks = {
            onProgress: (loaded, total) => {
                callbacks.postMessage({
                    type: 'loadProgress',
                    sourceId: source.id,
                    loaded,
                    total
                });
            },
            onError: (error) => {
                callbacks.postMessage({
                    type: 'loadError',
                    sourceId: source.id,
                    error: error.message
                });
            },
            enqueueMessage: (message) => {
                callbacks.postMessage(message);
            },
            getValidToken: callbacks.getValidToken
        };

        const options: PhotoSourceOptions = {
            maxPhotos: this.maxPhotosInArea
        };

        const loader = PhotoSourceFactory.createLoader(source, sourceCallbacks, options);
        this.loadingProcesses.set(source.id, loader);

        try {
            await loader.start(bounds);
        } catch (error) {
            if (!loader.isAborted() && !callbacks.shouldAbort(processId)) {
                console.error(`🢄PhotoOperations: Error loading source ${source.id}:`, error);
            }
        } finally {
            // Ensure loader is removed from active processes
            if (this.loadingProcesses.get(source.id) === loader) {
                this.loadingProcesses.delete(source.id);
            }
        }
    }

    async processCombinePhotos(
        processId: string,
        messageId: number,
        areaBounds: Bounds | null,
        sources: SourceConfig[],
        callbacks: OperationCallbacks
    ): Promise<void> {
        console.log(`🢄PhotoOperations: Processing combinePhotos (${processId})`);

        if (callbacks.shouldAbort(processId)) {
            console.log(`🢄PhotoOperations: CombinePhotos process ${processId} aborted before processing`);
            return;
        }

        // This operation combines and culls existing photos from all sources
        // No new loading - just processing current data using existing culling logic

        // The worker has the sophisticated culling logic (mergeAndCullPhotos)
        // Just trigger the standard photo update pipeline
        callbacks.sendPhotosInAreaUpdate();
        callbacks.sendPhotosInRangeUpdate();

        console.log(`🢄PhotoOperations: CombinePhotos processing complete (${processId}) - triggered photo updates`);
        callbacks.postMessage({
            type: 'processComplete',
            processId,
            processType: 'sourcesPhotosInArea',
            messageId
        });
    }

    private async simulateWork(
        processId: string,
        callbacks: OperationCallbacks,
        ms: number
    ): Promise<void> {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve(); // Always resolve to avoid hanging
            }, ms);
        });
    }
}

File: frontend/tauri-plugin-hillview/android/src/main/java/PhotoOperations.kt
package cz.hillview.plugin

import android.content.Context
import android.util.Log
import kotlinx.coroutines.*


/**
 * The three loaders behind the narrowest surface processArea/processConfig use.
 * PhotoOperations constructs the real ones by default; JVM unit tests hand in
 * fakes with controlled timing (see PhotoOperationsIncrementalTest).
 */
interface DeviceLoading {
    suspend fun load(source: SourceConfig, bounds: Bounds?, maxPhotos: Int, shouldAbort: () -> Boolean, picks: Set<String>): List<PhotoData>
}

interface StreamLoading {
    suspend fun load(
        source: SourceConfig, bounds: Bounds?, maxPhotos: Int, authToken: String?, shouldAbort: () -> Boolean,
        picks: Set<String>, queryOptionsJson: String?, onBatch: ((List<PhotoData>) -> Unit)?
    ): List<PhotoData>
}

interface PanoramaxLoading {
    suspend fun load(source: SourceConfig, bounds: Bounds?, maxPhotos: Int, shouldAbort: () -> Boolean, hillviewBackendUrl: String?, authToken: String?): List<PhotoData>
}

private class RealDeviceLoading(context: Context) : DeviceLoading {
    private val loader = DevicePhotoLoader(context)
    override suspend fun load(source: SourceConfig, bounds: Bounds?, maxPhotos: Int, shouldAbort: () -> Boolean, picks: Set<String>) =
        loader.loadPhotos(source, bounds, maxPhotos, shouldAbort, picks)
}

private class RealStreamLoading : StreamLoading {
    private val loader = StreamPhotoLoader()
    override suspend fun load(
        source: SourceConfig, bounds: Bounds?, maxPhotos: Int, authToken: String?, shouldAbort: () -> Boolean,
        picks: Set<String>, queryOptionsJson: String?, onBatch: ((List<PhotoData>) -> Unit)?
    ) = loader.loadPhotos(source, bounds, maxPhotos, authToken, shouldAbort, picks, queryOptionsJson, onBatch)
}

private class RealPanoramaxLoading : PanoramaxLoading {
    private val loader = PanoramaxPhotoLoader()
    override suspend fun load(source: SourceConfig, bounds: Bounds?, maxPhotos: Int, shouldAbort: () -> Boolean, hillviewBackendUrl: String?, authToken: String?) =
        loader.loadPhotos(source, bounds, maxPhotos, shouldAbort, hillviewBackendUrl, authToken)
}

/**
 * Photo Operations - Pure Business Logic
 *
 * Faithful translation of photoOperations.ts from TypeScript to Kotlin.
 * Handles the actual photo loading and processing operations.
 */
class PhotoOperations(
    private val context: Context,
    private val deviceLoader: DeviceLoading = RealDeviceLoading(context),
    private val streamLoader: StreamLoading = RealStreamLoading(),
    private val panoramaxLoader: PanoramaxLoading = RealPanoramaxLoading(),
    private val backendUrlProvider: () -> String? = { defaultHillviewBackendUrl(context) }
) {
    companion object {
        private const val TAG = "PhotoOperations"
        private const val doLog = false
        private const val MAX_PHOTOS_IN_AREA = 400  // Should match photoWorkerConstants.ts

        private fun defaultHillviewBackendUrl(context: Context): String? {
            // The frontend pushes its backend URL via set_backend_url (ExamplePlugin),
            // persisted to the shared "hillview_upload_prefs" prefs. Reused here so the
            // Panoramax loader can call /api/hidden/{photos,users} against it.
            return context
                .getSharedPreferences("hillview_upload_prefs", Context.MODE_PRIVATE)
                .getString("server_url", null)
        }
    }

    private fun hillviewBackendUrl(): String? = backendUrlProvider()
    private val sourceCache = mutableMapOf<String, SourceCache>()
    private var maxPhotosInArea: Int = MAX_PHOTOS_IN_AREA
    private var picks: Set<String> = emptySet()
    private var queryOptionsJson: String? = null  // Pre-serialized analysis filters

    /** Timestamp (System.currentTimeMillis()) recorded immediately before the most recent
     *  successful DevicePhotoLoader.loadPhotos() call. The frontend uses this to prune
     *  placeholder markers: if placeholder.savedAt < lastDeviceQueryStartedAt, the DB row
     *  was visible to the query, so the placeholder is redundant. */
    @Volatile
    var lastDeviceQueryStartedAt: Long? = null
        private set


    fun setPicks(newPicks: Set<String>) {
        picks = newPicks
    }

    fun setQueryOptionsJson(json: String?) {
        queryOptionsJson = json
    }

    /** Drop a single photo from the per-source cache so subsequent area loads
     *  don't resurrect it. Returns true if the photo existed. */
    fun removePhotoFromCache(photoId: String, sourceId: String): Boolean {
        val cache = sourceCache[sourceId] ?: return false
        val filtered = cache.photos.filterNot { it.id == photoId }
        if (filtered.size == cache.photos.size) return false
        sourceCache[sourceId] = cache.copy(photos = filtered)
        return true
    }

    /** Drop all photos by a given creator id from a source's cache. Returns the
     *  number of photos removed. */
    fun removeUserPhotosFromCache(userId: String, sourceId: String): Int {
        val cache = sourceCache[sourceId] ?: return 0
        val filtered = cache.photos.filterNot { it.creator?.id == userId }
        val removed = cache.photos.size - filtered.size
        if (removed > 0) sourceCache[sourceId] = cache.copy(photos = filtered)
        return removed
    }

    /**
     * Source cache for each source - matches TypeScript interface
     */
    private data class SourceCache(
        val photos: List<PhotoData>,
        val isComplete: Boolean, // true = no more photos available, false = partial load
        val cachedBounds: Bounds? = null // The geographic bounds that were completely cached
    )

    fun setMaxPhotosInArea(maxPhotos: Int) {
        maxPhotosInArea = maxPhotos
    }

    /**
     * Process config update - translation of processConfig from photoOperations.ts
     */
    suspend fun processConfig(
        processId: String,
        messageId: Int,
        config: ConfigData,
        shouldAbort: () -> Boolean,
        authTokenProvider: suspend () -> String?
    ): List<PhotoData> {
        if (doLog) Log.d(TAG, "PhotoOperations: Processing config update ($processId)")

        if (shouldAbort()) return emptyList()

        if (config.sources.isEmpty()) {
            if (doLog) Log.d(TAG, "PhotoOperations: PROCESSCONFIG: No sources in config ($processId)")
            return emptyList()
        }

        val sources = config.sources
        val allLoadedPhotos = mutableListOf<PhotoData>()

        // Get enabled sources
        val enabledSources = sources.filter { it.enabled }
        val enabledSourceIds = enabledSources.map { it.id }.toSet()

        if (doLog) Log.d(TAG, "PhotoOperations: PROCESSCONFIG: enabledSourceIds: ${enabledSourceIds.joinToString(", ")}")

        // Clear cache for disabled sources
        val sourcesToRemove = sourceCache.keys.filter { !enabledSourceIds.contains(it) }
        sourcesToRemove.forEach { sourceCache.remove(it) }

        // Process each enabled source
        for (source in enabledSources) {
            if (shouldAbort()) break

            try {
                if (doLog) Log.d(TAG, "PhotoOperations: Processing source ${source.id} of type ${source.type}")

                // Extract picks for this specific source
                // picks contain UIDs like "hillview-abc123", we need to extract "abc123" for the backend
                val sourcePrefix = "${source.id}-"
                val sourcePickIds = picks
                    .filter { it.startsWith(sourcePrefix) }
                    .map { it.substring(sourcePrefix.length) }
                    .toSet()

                val photos = when (source.type) {
                    "device" -> {
                        val queryStart = System.currentTimeMillis()
                        val result = deviceLoader.load(source, null, maxPhotosInArea, shouldAbort, sourcePickIds)
                        lastDeviceQueryStartedAt = queryStart
                        result
                    }
                    "stream" -> {
                        val authToken = authTokenProvider()
                        streamLoader.load(source, null, maxPhotosInArea, authToken, shouldAbort, sourcePickIds, queryOptionsJson, null)
                    }
                    "panoramax" -> {
                        val authToken = authTokenProvider()
                        panoramaxLoader.load(source, null, maxPhotosInArea, shouldAbort, hillviewBackendUrl(), authToken)
                    }
                    else -> {
                        Log.w(TAG, "PhotoOperations: Unknown source type: ${source.type}")
                        emptyList()
                    }
                }

                if (!shouldAbort()) {
                    allLoadedPhotos.addAll(photos)

                    // Cache the loaded photos
                    sourceCache[source.id] = SourceCache(
                        photos = photos,
                        isComplete = true, // For config processing, we consider this complete
                        cachedBounds = null
                    )

                    if (doLog) Log.d(TAG, "PhotoOperations: Loaded ${photos.size} photos from source ${source.id}")
                }

            } catch (error: Exception) {
                Log.e(TAG, "PhotoOperations: Error loading source ${source.id}: ${error.message}", error)
                // Continue with other sources
            }
        }

        if (doLog) Log.d(TAG, "PhotoOperations: PROCESSCONFIG: Config processing complete ($processId) - loaded ${allLoadedPhotos.size} photos")
        return allLoadedPhotos
    }

    /**
     * Process area update - translation of processArea from photoOperations.ts
     *
     * Every enabled source loads concurrently (supervisor semantics: one
     * source failing or stalling never touches the others), and each one
     * reports through [onSourcePhotos] as it lands — stream sources on every
     * SSE batch (the accumulated list so far), every source once more on
     * completion — so the caller can put markers on the map per source
     * instead of after the slowest one. The returned map is the settled set.
     */
    suspend fun processArea(
        processId: String,
        sources: List<SourceConfig>,
        bounds: Bounds,
        shouldAbort: () -> Boolean,
        authTokenProvider: suspend () -> String?,
        onSourceLoadingStatus: ((sourceId: String, isLoading: Boolean, progress: String?, error: String?) -> Unit)? = null,
        onSourcePhotos: ((sourceId: String, photos: List<PhotoData>) -> Unit)? = null
    ): Map<String, List<PhotoData>> {
        if (doLog) Log.d(TAG, "PhotoOperations: Processing area update ($processId) with ${sources.size} sources")

        val sourcesPhotosInArea = java.util.concurrent.ConcurrentHashMap<String, List<PhotoData>>()

        supervisorScope {
            for (source in sources.filter { it.enabled }) {
                if (shouldAbort()) break

                launch {
                    try {
                        // Send loading status for this individual source
                        onSourceLoadingStatus?.invoke(source.id, true, "Loading photos...", null)

                        if (doLog) Log.d(TAG, "PhotoOperations: Processing area for source ${source.id}")

                        // Check if we have cached data that covers this area
                        val cache = sourceCache[source.id]
                        val canUseCache = cache != null && cache.isComplete &&
                            cache.cachedBounds?.let { isAreaWithinCachedBounds(bounds, it) } == true

                        // Extract picks for this specific source
                        // picks contain UIDs like "hillview-abc123", we need to extract "abc123" for the backend
                        val sourcePrefix = "${source.id}-"
                        val sourcePickIds = picks
                            .filter { it.startsWith(sourcePrefix) }
                            .map { it.substring(sourcePrefix.length) }
                            .toSet()

                        val photos = if (canUseCache) {
                            if (doLog) Log.d(TAG, "PhotoOperations: Using cached data for ${source.id}")
                            filterPhotosByArea(cache!!.photos, bounds)
                        } else {
                            if (doLog) Log.d(TAG, "PhotoOperations: No cache for ${source.id}, performing bounded load (picks: ${sourcePickIds.size})")

                            when (source.type) {
                                "device" -> {
                                    val queryStart = System.currentTimeMillis()
                                    val result = deviceLoader.load(source, bounds, maxPhotosInArea, shouldAbort, sourcePickIds)
                                    lastDeviceQueryStartedAt = queryStart
                                    result
                                }
                                "stream" -> {
                                    val authToken = authTokenProvider()
                                    streamLoader.load(source, bounds, maxPhotosInArea, authToken, shouldAbort, sourcePickIds, queryOptionsJson) { batch ->
                                        // A batch of a superseded area is nobody's news.
                                        if (!shouldAbort()) onSourcePhotos?.invoke(source.id, batch)
                                    }
                                }
                                "panoramax" -> {
                                    val authToken = authTokenProvider()
                                    panoramaxLoader.load(source, bounds, maxPhotosInArea, shouldAbort, hillviewBackendUrl(), authToken)
                                }
                                else -> {
                                    Log.w(TAG, "PhotoOperations: Unknown source type: ${source.type}")
                                    emptyList()
                                }
                            }
                        }

                        if (!shouldAbort()) {
                            sourcesPhotosInArea[source.id] = photos
                            if (doLog) Log.d(TAG, "PhotoOperations: Area load complete for ${source.id}: ${photos.size} photos")

                            // Send completion status for this individual source
                            onSourceLoadingStatus?.invoke(source.id, false, "Loaded ${photos.size} photos", null)
                            onSourcePhotos?.invoke(source.id, photos)
                        }

                    } catch (cancelled: CancellationException) {
                        // Ours (a newer area superseded this one) — never report it as a source error.
                        throw cancelled
                    } catch (error: Exception) {
                        Log.e(TAG, "PhotoOperations: Error loading source ${source.id}: ${error.message}", error)

                        // Send error status for this individual source
                        onSourceLoadingStatus?.invoke(source.id, false, null, "Error: ${error.message}")

                        // Continue with other sources
                    }
                }
            }
        }
        /* The sequential loop this replaces (one source after another; the map
           only updated after the last one):
        for (source in sources.filter { it.enabled }) {
            if (shouldAbort()) break
            try {
                onSourceLoadingStatus?.invoke(source.id, true, "Loading photos...", null)
                ...
                val photos = if (canUseCache) { ... } else { when (source.type) { ... } }
                if (!shouldAbort()) {
                    sourcesPhotosInArea[source.id] = photos
                    onSourceLoadingStatus?.invoke(source.id, false, "Loaded ${photos.size} photos", null)
                }
            } catch (error: Exception) {
                Log.e(TAG, "PhotoOperations: Error loading source ${source.id}: ${error.message}", error)
                onSourceLoadingStatus?.invoke(source.id, false, null, "Error: ${error.message}")
            }
        }
        */

        if (doLog) Log.d(TAG, "PhotoOperations: Area processing complete ($processId) - ${sourcesPhotosInArea.values.sumOf { it.size }} photos in area")
        return sourcesPhotosInArea.toMap()
    }

    /**
     * Check if requested bounds is completely contained within cached bounds
     * Translation from photoOperations.ts
     */
    private fun isAreaWithinCachedBounds(requestedBounds: Bounds, cachedBounds: Bounds): Boolean {
        return requestedBounds.top_left.lat <= cachedBounds.top_left.lat &&
               requestedBounds.top_left.lng >= cachedBounds.top_left.lng &&
               requestedBounds.bottom_right.lat >= cachedBounds.bottom_right.lat &&
               requestedBounds.bottom_right.lng <= cachedBounds.bottom_right.lng
    }

    /**
     * Filter photos by area bounds
     * Translation from workerUtils.ts filterPhotosByArea
     */
    private fun filterPhotosByArea(photos: List<PhotoData>, bounds: Bounds): List<PhotoData> {
        val crossesAntimeridian = bounds.top_left.lng > bounds.bottom_right.lng
        return photos.filter { photo ->
            photo.coord.lat <= bounds.top_left.lat &&
            photo.coord.lat >= bounds.bottom_right.lat &&
            if (crossesAntimeridian) {
                photo.coord.lng >= bounds.top_left.lng || photo.coord.lng <= bounds.bottom_right.lng
            } else {
                photo.coord.lng >= bounds.top_left.lng && photo.coord.lng <= bounds.bottom_right.lng
            }
        }
    }

    /**
     * Clean up all resources
     */
    fun cleanup() {
        if (doLog) Log.d(TAG, "PhotoOperations: Cleaning up all resources")
        sourceCache.clear()
    }
}
